### PR TITLE
Feat release 4.1

### PR DESCRIPTION
**Pull request by barbieri**
Prepare release 4.1.0 fixing #52 and updating dependencies

### CODE REVIEW

File: tsconfig.json
Comment (cabelitos) [inline tsconfig.json:5]: you sure about this one? Not sure if this breaks old note versions like 12.

File: package.json
Comment (cabelitos) [inline package.json]: I was wondering. What do you think about this as dep and not a peer dep?

File: package.json
Comment (cabelitos) [inline package.json:56]: remember to delete the .env.example

Comment (cabelitos) [reply package.json:56]: Ignore me. it was deleted. lol.

**Review by cabelitos — DISMISSED** [review]

File: package.json
Comment (douglas-wang) [reply package.json]: Quick question, is there any plan to upgrade this to the latest version?
https://www.npmjs.com/package/graphql-upload

**Comment by bragovo** [comment]
Hello! When you planning to release this? We have a lot of pain about https://github.com/profusion/apollo-federation-file-upload/issues/52 :-D

File: package.json
Comment (barbieri) [reply package.json]: @cabelitos could be, I kept it as before, but I think it may be worth to have as dep as it's only use directly (unlike `@apollo/gateway` that is use by our users).

@douglas-wang it's already 15.0.1 at the devDependency, but I'll move this to a simple direct dependency.

File: tsconfig.json
Comment (barbieri) [reply tsconfig.json:5]: I'm not that sure, but comparing the output file it's the same.

As for node12, it's already not supported by Apollo itself, then we're 14 and up.

**Comment by barbieri** [comment]
> Hello! When you planning to release this? We have a lot of pain about #52 :-D

hi @bragovo, just sent a new try based on https://github.com/profusion/apollo-federation-file-upload/issues/52#issuecomment-1148973969, I'm waiting for @glasser to review so I can merge and release.

**Comment by barbieri** [comment]
@glasser, @bragovo would you mind confirming this works? I'll merge the fix proposed in #57 and release soon (today or tomorrow)

**Comment by bragovo** [comment]
@barbieri unfortunately my JS teammates still can't upload anything :/ I need some more time to check how to fix it.

**Comment by barbieri** [comment]
> @barbieri unfortunately my JS teammates still can't upload anything :/ I need some more time to check how to fix it.

weird, wasn't the `Headers` the problem? It should be using the plain `Object` now. Just check if the #57 pointed issue is not the problem (in other words: if you're relying on headers set at `willSendRequest`, we need that fix so it works.

**Comment by rostber** [comment]
That pull make problem. The `willSendRequest` method will call with `undefined` headers object (in original RemoteGraphQLDataSource headers always belongs to Headers and never `undefined`).

By the way, FileUploadDataSource class doesn't send `operationName`. Way to fix it:
```
    const operations = JSON.stringify({
      query: request.query,
      variables,
      operationName:
        'incomingRequestContext' in args
          ? args.incomingRequestContext.request.operationName
          : args.request.operationName,
    });
    ```

**Comment by TheBit** [comment]
Hi, any news on this?
We can't go live without upload :(

**Comment by sfounder** [comment]
Hello. Is it so critical to merge? I see few developers are blocked this MR. We also can't go live without upload. Could you merge this MR and then refactor code? Thanks!

**Comment by paulpdaniels** [comment]
Any help that the maintainers need here? I know we all get busy, so anything I can do to help move this along?

**Comment by glasser** [comment]
Hi, I found a bit of time to look at this but I'm not sure what I'm being asked to review. It looks like you're working around the `Fetcher` type declaration with an `as` typecast, which you need because you're trying to put a FormData as the body which `Fetcher` doesn't expect. But [as I said](https://github.com/profusion/apollo-federation-file-upload/issues/52#issuecomment-1148973969) I don't think that arbitrary Fetch implementations all handle FormData in the same way. So doing it this way is essentially locking your API in to only working with Fetch implementations that happen to work with the FormData implementation you chose. To me this seems likely to lead to confusion. My [recommendation continues to be](https://github.com/profusion/apollo-federation-file-upload/issues/52#issuecomment-1148973969) that if this package depends on the ability to send FormData (as implemented by the `form-data` npm package) via a Fetch API, then it should just choose a particular implementation of Fetch that definitely supports the `form-data` package and use it directly instead of allowing the fetcher to be overridden with an arbitrary implementation that might not work.

**Comment by glasser** [comment]
(That said, if you're OK with the idea that your package will be fragile and could break if a user passes in a Fetcher implementation that doesn't support FormData from `form-data` as a body, then this is probably fine? I'm not an expert on the details of your project but if it works and passes tests then it's probably OK?)

File: lib/FileUploadDataSource.ts
Comment (DDDKnightmare) [inline lib/FileUploadDataSource.ts:12]: Considering what has been suggested [here](https://github.com/profusion/apollo-federation-file-upload/pull/55#issuecomment-1262791533), wouldn't it be better to add `make-fetch-happen` as direct dependency and force it to be used instead of using the default from apollo? (it uses `make-fetch-happen` at the moment, but this can change in future versions).

**Review by DDDKnightmare — APPROVED** [review]

**Review by douglas-wang — APPROVED** [review]

**Comment by TheBit** [comment]
@cabelitos sorry for disturbing, but this is waiting for half a year, are there any plans for this?

**Comment by BobbiSixkiller** [comment]
Is this going to be released any time soon please ?

**Comment by oliveirarleo** [comment]
Hello @TheBit and @BobbiSixkiller. We cannot expect help from @cabelitos anymore, but our team at ProFUSION will put some hands here to help you guys out right at our next sprint. We now have officially including this task in our planning. That means we'll start working (and revise the work done here) as of 30th of January and we expect to solve the problem and merge between 2 and 4 weeks.

File: lib/FileUploadDataSource.ts
Comment (g7fernandes) [inline lib/FileUploadDataSource.ts:209]: test covering this here https://github.com/profusion/apollo-federation-file-upload/pull/67

**Review by g7fernandes — APPROVED** [review]

**Comment by oliveirarleo** [comment]
:tada:

**Pull request merged**
